### PR TITLE
SSL support for downstream connections

### PR DESCRIPTION
**Pull request by abc3**
This PR introduces support for establishing secure connections between Supavisor and its clients.

To utilize this feature, the `GLOBAL_DOWNSTREAM_CERT_PATH` and `GLOBAL_DOWNSTREAM_KEY_PATH` environment variables need to be declared, pointing to the corresponding files.

Upon receiving an SSL request, Supavisor checks if these variables are defined. If they are, it sends an `S` response and initiates a TLS handshake procedure. If not, Supavisor sends an `N` message over the TCP socket to the client, indicating that the server doesn't support SSL communication. The client can then choose to either maintain an unsecured connection or terminate it

### CODE REVIEW

File: config/runtime.exs [hunk 1]
```diff
@@ -102,8 +103,32 @@ upstream_ca =
     File.read!(path)
     |> H.cert_to_bin()
     |> case do
-      {:ok, bin} -> bin
-      {:error, _} -> raise "There is no valid certificate in $GLOBAL_UPSTREAM_CA_PATH"
+      {:ok, bin} ->
+        Logger.warning("Loaded upstream CA from $GLOBAL_UPSTREAM_CA_PATH")
```
[the remark this answers is not in the document]
Comment (abc3) [reply config/runtime.exs:107]: The main thing why I use a warning just to highlight it in the console during development :D
Let me know if this seems odd

Screenshot 2023-07-11 at 14 32 20

[the remark this answers is not in the document]
Comment (wojtekmach) [reply config/runtime.exs:107]: If it's just for dev, I'd consider Logger.debug. I think for warnings and errors you may want to have some monitoring eventually and then these entries would be noise.

[the remark this answers is not in the document]
Comment (josevalim) [reply config/runtime.exs:107]: I believe you can also pass ansi_color in the metadata with one of IO.ANSI functions result! :)

File: lib/supavisor/client_handler.ex [hunk 2]
```diff
@@ -48,29 +53,53 @@ defmodule Supavisor.ClientHandler do
...
+    sock_send(data.sock, "HTTP/1.1 204 OK\r\n\r\n")
     {:stop, :normal, data}
   end
 
-  def handle_event(:info, {:tcp, _, <<_::64>>}, :exchange, data) do
-    Logger.warn("Client is trying to connect with SSL")
-    # TODO: implement SSL negotiation
-    # SSL negotiation, S/N/Error
-    :gen_tcp.send(data.socket, "N")
+  def handle_event(:info, {:tcp, _, <<_::64>>}, :exchange, %{sock: {_, tcp_sock}} = data) do
```
[the remark this answers is not in the document]
Comment (abc3) [reply lib/supavisor/client_handler.ex:70]: Yeah, good call 👍

File: config/runtime.exs (hunk 1 above)
[the remark this answers is not in the document]
Comment (abc3) [reply config/runtime.exs:107]: @wojtekmach, only a different color is needed for debugging

@josevalim, thanks for the advice 🔥

**Comment by abc3** [comment]
@wojtekmach could you take a look one more time?

**Review by wojtekmach — COMMENTED** [review]
LGTM!

**Review by delgado3d — APPROVED** [review]

**Pull request merged**